src/bounds.py

```diff
@@ -11,12 +11,30 @@ def lower_bound(x: list[int], v: int) -> int:
 
     If all values in x are smaller than v, return len(x).
     """
-    return 0  # FIXME: Obviously the answer isn't always 0
+    start = 0 # start of search interval
+    end = len(x) # end of search interval
 
+    while start < end:
+        middle = (start+end)//2 # middle of interval
+        if  v <= x[middle]:
+            end = middle
+        else:
+            start = middle + 1
+    return start
+ 
 
 def upper_bound(x: list[int], v: int) -> int:
     """Get the index of the upper bound of v in x.
 
     If all values in x are smaller than v, return len(x).
     """
-    return 0  # FIXME: Obviously the answer isn't always 0
+    start = 0 # start of search interval
+    end = len(x) # end of search interval
+
+    while start < end:
+        middle = (start+end)//2 # middel of interval
+        if  v < x[middle]:
+            end = middle
+        else:
+            start = middle + 1
+    return start
```

Comment:
You can also exploit that when we use integers, `upper_bound(x,v) == lower_bound(x,v+1)`